### PR TITLE
Make array mutable in combinatorics.jl

### LINKED ISSUE #47847: zero should be mutable in `invperm`
https://github.com/JuliaLang/julia/blob/6bfc6ac7187f04cc2554425ecc8a2fa42b4f1af8/base/combinatorics.jl#L279-L285

On line 279, this should be `fill!(similar(a), zero(eltype(a)))`, as `zero(a)` may not be mutable (e.g. for `StaticArray`s).

### PR DESCRIPTION
**Pull request by shogohida**
Signed-off-by: Shogo Hida <shogo.hida@gmail.com>

Fixes #47847

### CODE REVIEW

File: base/combinatorics.jl [hunk 1]
```diff
@@ -276,7 +276,7 @@ julia> B[invperm(v)]
 """
 function invperm(a::AbstractVector)
     require_one_based_indexing(a)
-    b = zero(a) # similar vector of zeros
+    fill!(similar(a), zero(eltype(a))) # similar vector of zeros
```
Comment (simeonschaub) [inline base/combinatorics.jl:279]: We're filling each entry in `b` anyways, so we don't actually need to initialize it with zeros.

```suggestion
    b = similar(a)
```

File: base/combinatorics.jl [hunk 2]
```diff
@@ -276,7 +276,7 @@ julia> B[invperm(v)]
...
     n = length(a)
     @inbounds for (i, j) in enumerate(a)
         ((1 <= j <= n) && b[j] == 0) ||
```
Comment (simeonschaub) [inline base/combinatorics.jl:282]:
```suggestion
        (1 <= j <= n) ||
```

**Review by simeonschaub — COMMENTED** [review]
Thanks for the PR! Just a few comments. This will also need a test. I think the best strategy for that would be to use the `SizedArray` type in `test/testhelpers/SizedArrays.jl`, defining a method for `similar` returning a `SizedArray` to avoid hitting the default fallback, and then testing that `invperm` actually returns a `SizedArray` here.

File: base/combinatorics.jl (hunk 2 above)
Comment (fredrikekre) [reply base/combinatorics.jl:282]: This is to make sure `b[j]` haven't been visited before (there is no `isperm(a)`). With your suggestions:
```
julia> invperm([1, 1, 1])
3-element Vector{Int64}:
                   3
 7021782956690391410     #  garbage from similar(...)
                   0     #  garbage from similar(...)
```

File: base/combinatorics.jl (hunk 1 above)
Comment (jishnub) [reply base/combinatorics.jl:279]: This is necessary, as the zero check is carried out in there loop

File: base/combinatorics.jl (hunk 2 above)
Comment (simeonschaub) [reply base/combinatorics.jl:282]: Ah, I see! I didn't realized these weren't just visited in order, ignore my comment

**Comment by jishnub** [comment]
Maybe we should have a test for which `zero(a)` is immutable to prevent regressions? A `SizedArray` has a mutable zero.

File: base/combinatorics.jl [hunk 3]
```diff
@@ -276,10 +276,10 @@ julia> B[invperm(v)]
 """
 function invperm(a::AbstractVector)
     require_one_based_indexing(a)
-    b = zero(a) # similar vector of zeros
+    b = fill!(similar(a), zero(eltype(a))) # similar vector of zeros
     n = length(a)
     @inbounds for (i, j) in enumerate(a)
-        ((1 <= j <= n) && b[j] == 0) ||
+        (1 <= j <= n) ||
```
Comment (petvana) [inline base/combinatorics.jl:282]: Why do you remove `b[j] == 0` check? I'm surprised this is not covered by CI.

Comment (shogohida) [reply base/combinatorics.jl]: Oh, sorry about that. I misunderstood [this comment](https://github.com/JuliaLang/julia/pull/47855#discussion_r1044793858) not following the discussion.

**Comment by shogohida** [comment]
Hi all, I've got a question. Will the test be like this?

```
test_array = similar([SizedArray{(2,2)}([1 2;3 4])])
@test typeof(invperm(test_array)) == SizedArray
```

**Comment by jishnub** [comment]
Hey @shogohida, sorry for the extremely long delay in replying (1 year!). If you're still interested, you may use the `ImmutableArray` test helper

**Pull request closed without merging**